File: src/main/java/dev/lazurite/rayon/impl/util/Frame.java
package dev.lazurite.rayon.impl.util;

import com.jme3.math.Quaternion;
import com.jme3.math.Vector3f;
import dev.lazurite.rayon.api.PhysicsElement;
import dev.lazurite.rayon.utils.math.MathShit;

/**
 * A {@link Frame} can be used for interpolation on the render thread.
 * {@link Frame}s are stored in {@link PhysicsElement}s and are updated
 * each tick.
 */
public class Frame {
    private Vector3f prevLocation;
    private Vector3f tickLocation;
    private Quaternion prevRotation;
    private Quaternion tickRotation;

    public Frame() {
        this(new Vector3f(), new Quaternion());
    }

    public Frame(Vector3f location, Quaternion rotation) {
        this.set(location, location, rotation, rotation);
    }

    public void set(Vector3f prevLocation, Vector3f tickLocation, Quaternion prevRotation, Quaternion tickRotation) {
        this.prevLocation = prevLocation;
        this.tickLocation = tickLocation;
        this.prevRotation = prevRotation;
        this.tickRotation = tickRotation;
    }

    public void from(Frame frame) {
        this.set(frame.prevLocation, frame.tickLocation, frame.prevRotation, frame.tickRotation);
    }

    public void from(Frame prevFrame, Vector3f tickLocation, Quaternion tickRotation) {
        this.set(prevFrame.tickLocation, tickLocation, prevFrame.tickRotation, tickRotation);
    }

    public Vector3f getLocation(Vector3f store, float tickDelta) {
        Vector3f lerp = MathShit.lerp(this.prevLocation, this.tickLocation, tickDelta);
        return store.set(lerp);
    }

    public Quaternion getRotation(Quaternion store, float tickDelta) {
        Quaternion slerp = MathShit.slerp(this.prevRotation, this.tickRotation, tickDelta);
        return store.set(slerp);
    }

    public Vector3f getLocationDelta(Vector3f store) {
        return store.set(tickLocation.subtract(prevLocation));
    }

    public Vector3f getRotationDelta(Vector3f store) {
        final var tickAngles = MathShit.toEulerAngles(this.tickRotation);
        final var prevAngles = MathShit.toEulerAngles(this.prevRotation);
        tickAngles.subtractLocal(prevAngles);
        return store.set(tickAngles);
    }
}

File: src/main/java/ru/melonhell/bulletphysics/bullet/collision/body/element/Frame.kt
package ru.melonhell.bulletphysics.bullet.collision.body.element

import com.jme3.math.Quaternion
import com.jme3.math.Vector3f
import ru.melonhell.bulletphysics.utils.math.MathShit.lerp
import ru.melonhell.bulletphysics.utils.math.MathShit.slerp
import ru.melonhell.bulletphysics.utils.math.MathShit.toEulerAngles

/**
 * A [Frame] can be used for interpolation on the render thread.
 * [Frame]s are stored in {link PhysicsElement}s and are updated
 * each tick.
 */
class Frame(
    location: Vector3f = Vector3f(),
    rotation: Quaternion = Quaternion()
) {
    private var prevLocation: Vector3f = Vector3f()
    private var tickLocation: Vector3f = Vector3f()
    private var prevRotation: Quaternion = Quaternion()
    private var tickRotation: Quaternion = Quaternion()

    init {
        this[location, location, rotation] = rotation
    }

    operator fun set(
        prevLocation: Vector3f,
        tickLocation: Vector3f,
        prevRotation: Quaternion,
        tickRotation: Quaternion
    ) {
        this.prevLocation = prevLocation
        this.tickLocation = tickLocation
        this.prevRotation = prevRotation
        this.tickRotation = tickRotation
    }

    fun from(frame: Frame) {
        this[frame.prevLocation, frame.tickLocation, frame.prevRotation] = frame.tickRotation
    }

    fun from(prevFrame: Frame, tickLocation: Vector3f, tickRotation: Quaternion) {
        this[prevFrame.tickLocation, tickLocation, prevFrame.tickRotation] = tickRotation
    }

    fun getLocation(store: Vector3f, tickDelta: Float): Vector3f {
        val lerp = lerp(prevLocation, tickLocation, tickDelta)
        return store.set(lerp)
    }

    fun getRotation(store: Quaternion, tickDelta: Float): Quaternion {
        val slerp = slerp(prevRotation, tickRotation, tickDelta)
        return store.set(slerp)
    }

    fun getLocationDelta(store: Vector3f): Vector3f {
        return store.set(tickLocation.subtract(prevLocation))
    }

    fun getRotationDelta(store: Vector3f): Vector3f {
        val tickAngles = toEulerAngles(tickRotation)
        val prevAngles = toEulerAngles(prevRotation)
        tickAngles.subtractLocal(prevAngles)
        return store.set(tickAngles)
    }
}